soc/intel_adsp: Update cavs-fw.py loader for cAVS 2.5 devices

andyross opened the pull request:
Compatibility work (actually more or less a rewrite of large chunks)
to get the loader script working on a Tiger Lake cAVS 2.5 device.

Also one major fix (required for timer interrupts to arrive) and a few tweaks to get most tests running while I work on SMP.

Commit:
soc/intel_adsp: Increase init priority of trace windows

These windows control host visibility of the trace output buffer.  The
buffer itself is writable memory always, but until we get to the
register init the host can't see them.  Since they contain
printk/logging output, they REALLY need to be initialized earlier than
anything else.

Also remove a rogue memset of the trace buffer.  That buffer is
already being initialized in a lazy-evaluated way by the trace output
code, and blowing it away here has the effect of forgetting anything
earlier code was trying to log!

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Update cavs-fw.py loader for cAVS 2.5 devices

Add a loader script variant for Tiger Lake (cAVS 2.5) devices, which
have very slightly different loading behavior from older 1.5 DSPs.

This is added as a "-v25.py" script, and the original has been renamed
to cavs-fw-v15.py.  Note that there is no good reason except schedule
pressure that these are not the same script, I just wasn't able to
make a single script work compatibly in the time available.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

nashif commented on samples/hello_world/src/main.c:
```diff
@@ -7,7 +7,23 @@
 {
 	printk("Hello World! %s\n", CONFIG_BOARD);
+
+	uint32_t *capregs = (void *)0x1000;
```
debug code?

nashif commented on tests/kernel/common/src/errno.c:
```diff
@@ -47,15 +47,22 @@ static void errno_thread(void *_n, void *_my_errno, void *_unused)
 	int n = POINTER_TO_INT(_n);
 	int my_errno = POINTER_TO_INT(_my_errno);
 
+	printk("ANDY %s:%d\n", __func__, __LINE__);
```
debug code

andyross replied:
Yeah, stray "-a" in the git commit sucked in my whole tree and not the file I was trying to push.  :)  Should be fixed.

andyross commented:
FWIW: to I'm sure no one's surprise this has broken cavs15 loading somewhere.  That's not a huge deal as for that platform we have a working fallback in the older scripting (which I think Zephyr QA is still using) and it's no longer a focus platform for Zephyr development (we already made all the tests pass there and just need to fix regressions).  But it's surely fixable.

nashif commented:
> That's not a huge deal as for that platform we have a working fallback in the older scripting (which I think Zephyr QA is still using) and it's no longer a focus platform for Zephyr development (we already made all the tests pass there and just need to fix regressions). But it's surely fixable.

Can we please keep this working for 1.5 by either creating a different script for 2.5 and leaving 1.5 intact or by providing compatibility in the script itself?

Commit:
soc/intel_adsp: Fix timing/clock register ownership on cAVS 1.8+

The wall clock timer is not (per documentation) part of the
"timestamping" register set on the DSP.  And its counter and
comparator registers work fine always.  But if the DSP isn't set as
the "owner" of the timestamp hardware, wall clock interrupts never
arrive.

Also grab the PLL ownership too, because SOF already does anyway.
While we don't have a dynamic clock driver yet, we will surely want
one soon and will needt this.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
board/intel_adsp/cavs_v25: Use only one CPU for now

SMP bringup is still ongoing, so set the default here to 1 so that
uniprocessor testing can proceed (seems like most everyghing passes
already).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

lgirdwood commented:
> > That's not a huge deal as for that platform we have a working fallback in the older scripting (which I think Zephyr QA is still using) and it's no longer a focus platform for Zephyr development (we already made all the tests pass there and just need to fix regressions). But it's surely fixable.
> 
> Can we please keep this working for 1.5 by either creating a different script for 2.5 and leaving 1.5 intact or by providing compatibility in the script itself?

Ack, this needs to work for all as it's needed for the removal of xtos in SOF (alongside BYT and BDW SoC code)..

Commit:
boards/intel_adsp: Update cavs-fw.py, robustify vs. Linux PM

Add support for the extended manifest blocks that recent versions of
rimage are including as a prefix on the firmware blob.

Also include some PCI runtime PM tricks to reliably enable the DSP
device on kernels that have turned off the DSP device automatically
(e.g. systems that have PM enabled by default but where the SOF driver
is not loaded -- chromebooks work this way, potentially other
distros).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Yeah, agreed.  I've split the two scripts into separate files for now (again, there's no good reason we can't have one, I just don't have time and this stuff is really annoying to deal with -- changing one thing breaks stuff in weird ways).

Chased down the timer interrupt bug, that's included here, along with an SMP disable that unblocks testing while I sort out bringing the second CPU up (status: working as far as the ADSPCS and PWRSTS registers are concerned -- the cores seem to be alive and powered, but they aren't receiving the IDC interrupt to kick them out of the ROM and into our code, currently digging around in the IDC and interrupt masking layers and comparing to SOF trying to figure out what's broken).

Commit:
boards/intel_adsp_cavs25: Add board documentation

The default development environment for this target is in fact a
consumer Chromebook available to anyone.  So it deserves public
documentation.

Note that this spends considerably more time explaining the details of
how to configure a chromebook as a Zephyr development platform than it
does the process for building and running Zephyr (which is really very
conventional).  Chromebooks, which allow user-signed audio firmware,
are a great boon to SOF development.  But they were never intended as
developer devices themselves.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

kv2019i commented on boards/xtensa/intel_adsp_cavs25/doc/index.rst:
```diff
@@ -0,0 +1,409 @@
+
+On your build host, grab a copy of the ChromeOS kernel tree.  The
+shipping device is using a 5.4 kernel, but the 5.10 tree works for me
+and seems to have been backporting upstream drivers such that its main
```
You could add a remark to explain why this is needed. I guess the only mandatory req is the hugetbl support.

andyross replied:
That's right.  And in fact we don't strictly need hugetlbfs.  Transparent huge pages would work too (and are supported by default in chromeos kernels), I was just a little confused about the rules for when they're provided vs. when you get 4k pages as a fallback, and wasn't completely sure how to detect them from the /proc pagemap file.  The filesystem has simpler semantics.

And even then, we don't actually NEED the 2M page.  With only 4k pages, you can set up a DMA stream with 1M of buffer, which is enough to store (hopefully) any firmware we want to push.  But setting up scatter gather DMA from userspace seemed too difficult vs. just using a known-contiguous region. :)

andyross replied:
But it's true: without the hugetlbfs need, Zephyr itself probably would be fine with the factory 5.4 kernel.  I guess I'm assuming that all these devices are going to want to be running current SOF kernel drivers at some point, since at the end of the day that's the only app we truly need to deploy.

kv2019i commented on boards/xtensa/intel_adsp_cavs25/doc/index.rst:
```diff
@@ -0,0 +1,409 @@
+.. code-block:: console
+
+    $ scp kernel.img root@cros:/tmp
+    $ ssh root@cros dd if=/tmp/kernel.img of=/dev/nvme0n1p4
```
Updating modules?

andyross replied:
Happened a few steps above right after make MODULES_INSTALL.  ChromeOS has no initrd, you just drop the modules in the root filesystem.  Maybe it would be clearer to split the step and do that here, though.

kv2019i reviewed (changes requested):
Excellent, thanks @andyross ! A few minor comments.

kv2019i reviewed (approved):
Thanks for the response. I think these update to docs can done incrementally as well.

andyross commented:
Will update.  Unfortunately it's gone and broken on my board and I need to untangle what I did...

andyross commented:
Close this.  All the changes got merged into the more recent MP pull request at #37647

andyross commented:
Oops, actually close this.
